### DISCUSSION TITLE
Efficient Data Passing from Parent to Child Server Component

### DISCUSSION THREAD

**Discussion by palak-temp** [vercel/next.js#53513] [App Router] [answered]
### Goals

I aim to efficiently pass data from the parent server component to the child server component. I comprehend the concept of deduping, but I'm uncertain about handling dynamic routes since the params are accessible only to the base (parent) component and not the child component. 

## Demo
`blog/[id]/page.tsx`

```tsx
import React from 'react'
import BlogBody from './BlogBody'

export interface Blog {
  userId: number
  id: number
  title: string
  body: string
}

export async function getBlog({ params }: { params: { id: string } }) {
  const res = await fetch(`https://jsonplaceholder.typicode.com/posts/${params.id}`, { cache: 'no-store' })
  // The return value is *not* serialized
  // You can return Date, Map, Set, etc.

  if (!res.ok) {
    // This will activate the closest `error.js` Error Boundary
    throw new Error('Failed to fetch data')
  }

  return res.json()
}

async function Blog({ params: { id = "0" } }: { params: { id: string } }) {
  const blog: Blog = await getBlog({ params: { id } })
  return (
    <div>
      <h1 className="text-5xl">{blog.title}</h1>
      <BlogBody />
    </div>
  )
}

export default Blog
```

`BlogBody.tsx`

```
import React from 'react'

async function BlogBody() {

  // TODO: How to fetch data like parent on nested level
  const blog = {}

  return (
    <>
      <pre>{JSON.stringify(blog, undefined, 2)}</pre>
    </>
  )
}

export default BlogBody
```

### Non-Goals

- Deduping won't be effective in this case since the child server component doesn't have access to the params.
- Utilizing parallel routes won't significantly improve the user experience, just like the issue with props drilling

### Background

Props drilling and parallel routes are alternative approaches that may not effectively serve the Developer Experience Developer's

### Proposal

Requesting for the effective way.

**Comment by rinald-ta** [comment, marked as answer]
@palaklive While there isn't any documented way to pass props down to nested server components, the canary version of React (used by the app router) does export `createServerContext`, which is the equivalent of `createContext`, but specifically for server components. Keep in mind that this is an undocumented API and it's probably not intended to be used in regular code just yet.

Expanding on your example:

`blog/[id]/page.tsx`

```tsx
import { createServerContext } from 'react'
import BlogBody from './BlogBody'

export interface Blog {
  userId: number
  id: number
  title: string
  body: string
}

// create a params context
export const ParamsContext = createServerContext('params', {})

export async function getBlog(id: string) {
  const res = await fetch(`https://jsonplaceholder.typicode.com/posts/${id}`, {
    cache: 'no-store',
  })
  // The return value is *not* serialized
  // You can return Date, Map, Set, etc.

  if (!res.ok) {
    // This will activate the closest `error.js` Error Boundary
    throw new Error('Failed to fetch data')
  }

  return res.json()
}

async function Blog({ params: { id = '0' } }: { params: { id: string } }) {
  const blog: Blog = await getBlog(id)

  return (
    <ParamsContext.Provider value={{ id }}>
      <div>
        <h1 className="text-5xl">{blog.title}</h1>
        <BlogBody />
      </div>
    </ParamsContext.Provider>
  )
}

export default Blog

```

`BlogBody.tsx`

```tsx
import { use, Usable } from 'react'
import { ParamsContext, getBlog } from './page'

async function BlogBody() {
  const { id } = use(ParamsContext as Usable<any>)

  const blog = await getBlog(id)

  return <pre>{JSON.stringify(blog, undefined, 2)}</pre>
}

export default BlogBody
```

Hope this helps. If you need another solution, you can take a look at this similar discussion I had opened some time ago: #51818

**Reply by palak-temp** [reply]
Hi @rinald-ta 👋

Thank you for your response! I appreciate your help in clarifying the concept of handling dynamic routes in the context of passing data from the parent server component to the child server component.

You mentioned the canary version of React and the use of `createServerContext.` This sounds promising. However, I'd like to proceed with caution. Could you please provide me with an official documentation reference or any other resources where I can learn more about `createServerContext` and its usage in server components?

**Reply by rinald-ta** [reply]
There isn't any official documentation related to React Canaries, at least that I'm aware of. I only recently learned about `createServerContext` through a playground repo by one of the maintainers: https://github.com/ijjk/context-playground

You can also check out this blog post to learn more about the React Canary release channel: https://react.dev/blog/2023/05/03/react-canaries

**Comment by palak-temp** [comment]
@rinald-ta has proven to be incredibly useful, and I highly recommend it. Meanwhile, while browsing Stack Overflow, I stumbled upon this gem that I believe deserves some attention. Let me share it here to give it a bit more exposure.

---
`server-context.tsx`
```tsx
import 'server-only'
import { cache } from 'react'

/**
 * Creates a server context with a default value.
 * @param defaultValue The default value for the server context.
 * @returns A tuple of functions: [getValue, setValue].
 */
export default function createServerContext<T>(defaultValue: T): [() => T, (value: T) => void] {
  const getRef = cache(() => ({ current: defaultValue }))

  const getValue = (): T => getRef().current

  const setValue = (value: T): void => {
    getRef().current = value
  }

  return [getValue, setValue]
}
```

`blog/[id]/page.tsx`

```tsx
import createServerContext from '@/server-context'
import BlogBody from './BlogBody'

export interface Blog {
  userId: number
  id: number
  title: string
  body: string
}

export const idContext = createServerContext("0")

export async function getBlog(id: string) {
  const res = await fetch(`https://jsonplaceholder.typicode.com/posts/${id}`, {
    cache: 'no-store',
  })
  // The return value is *not* serialized
  // You can return Date, Map, Set, etc.

  if (!res.ok) {
    // This will activate the closest `error.js` Error Boundary
    throw new Error('Failed to fetch data')
  }

  return res.json()
}

async function Blog({ params: { id = '0' } }: { params: { id: string } }) {
  const blog: Blog = await getBlog(id)
  const [_, setId] = idContext
  setId(id)

  return (
    <div>
      <h1 className="text-5xl">{blog.title}</h1>
      <BlogBody />
    </div>
  )
}

export default Blog
```

`BlogBody.tsx`

```tsx
import { idContext, getBlog } from './page'

async function BlogBody() {
  const [id] = idContext
  const blog = await getBlog(id())

  return <pre>{JSON.stringify(blog, undefined, 2)}</pre>
}

export default BlogBody
```

### Note
---
The only concern revolves around how these variables utilize server memory and when they are eventually destroyed. 😅

**Reply by palak-temp** [reply]
A community has created an npm package to tackle the same issue. You can find it at this reference: [server-only-context](https://www.npmjs.com/package/server-only-context)
